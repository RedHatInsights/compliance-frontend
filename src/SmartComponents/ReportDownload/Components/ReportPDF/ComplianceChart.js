import React from 'react';
import propTypes from 'prop-types';
import { paletteColors } from '../../../../constants';
// eslint-disable-next-line rulesdir/disallow-fec-relative-imports
import { Chart } from '@redhat-cloud-services/frontend-components-pdf-generator/dist/esm/index';
import { fixedPercentage } from 'Utilities/TextHelper';

// TODO Legend table style need to be disablable
const ComplianceChart = ({
  policy: { percentCompliant = 0 },
  compliantSystemCount,
  nonCompliantSystemCount,
  unsupportedSystemCount,
  nonReportingSystemCount,
}) => {
  const compliantSystemsChartData = [
    {
      x: `${compliantSystemCount} systems compliant`,
      y: compliantSystemCount,
    },
    {
      x: `${nonCompliantSystemCount} systems non-compliant`,
      y: nonCompliantSystemCount,
    },
    ...(unsupportedSystemCount > 0
      ? [
          {
            x: `${unsupportedSystemCount} systems not supported`,
            y: unsupportedSystemCount,
            color: paletteColors.gold300,
          },
        ]
      : []),
    ...(nonReportingSystemCount > 0
      ? [
          {
            x: `${nonReportingSystemCount} systems never reported`,
            y: nonReportingSystemCount,
          },
        ]
      : []),
  ];
  const compliancePercentage = fixedPercentage(percentCompliant);

  return (
    <Chart
      legendHeader={''}
      chartType="donut"
      subTitle="Compliant"
      colorSchema="multi"
      title={compliancePercentage}
      data={compliantSystemsChartData}
    />
  );
};

ComplianceChart.propTypes = {
  policy: propTypes.object,
  compliantSystemCount: propTypes.number,
  nonCompliantSystemCount: propTypes.number,
  unsupportedSystemCount: propTypes.number,
  nonReportingSystemCount: propTypes.number,
};

export default ComplianceChart;
