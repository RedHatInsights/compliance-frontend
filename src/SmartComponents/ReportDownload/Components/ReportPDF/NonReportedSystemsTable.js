import React from 'react';
import propTypes from 'prop-types';
// eslint-disable-next-line rulesdir/disallow-fec-relative-imports
import { Table } from '@redhat-cloud-services/frontend-components-pdf-generator/dist/esm/index';

const NonReportedSystemsTable = ({ systems }) => {
  const headerRow = ['System name', 'OS'];
  const rows = systems.map((system) => [
    system.name,
    `RHEL ${system.osMajorVersion}.${system.osMinorVersion}`,
  ]);

  return <Table withHeader rows={[headerRow, ...rows]} />;
};

NonReportedSystemsTable.propTypes = {
  systems: propTypes.array,
  ssgFinder: propTypes.func,
};

export default NonReportedSystemsTable;
