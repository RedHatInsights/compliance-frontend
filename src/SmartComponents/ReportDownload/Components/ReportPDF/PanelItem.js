import React from 'react';
import propTypes from 'prop-types';
import { Text, View } from '@react-pdf/renderer';

// TODO: Move into pdf-generator components as option to render title at the bottom
const PanelItem = (props) => {
  const { children, title } = props;
  return (
    <View
      style={{
        justifyContent: 'space-evenly',
        width: '24%',
        paddingRight: '1%',
      }}
    >
      <View>
        <Text style={{ color: '#C9190B', fontSize: 20 }}>{children}</Text>
      </View>
      <Text>{title}</Text>
    </View>
  );
};

PanelItem.propTypes = {
  children: propTypes.node,
  title: propTypes.string,
};

export default PanelItem;
