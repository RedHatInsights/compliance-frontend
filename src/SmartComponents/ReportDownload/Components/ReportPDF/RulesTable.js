import React from 'react';
import propTypes from 'prop-types';
// eslint-disable-next-line rulesdir/disallow-fec-relative-imports
import { Table } from '@redhat-cloud-services/frontend-components-pdf-generator/dist/esm/index';

const identifierLabel = ({ identifier }) => (identifier || '{}').label || '';

const capitalize = (str = '') => str.charAt(0).toUpperCase() + str.slice(1);

const RulesTable = ({ rules }) => {
  const headerRow = ['Rule name', 'ID', 'Severity', 'Failed systems'];
  const failedRuleRows = rules.slice(0, 10).map((rule) => [
    rule.title,
    identifierLabel(rule),
    // TODO: Add icon svg for severity
    capitalize(rule?.severity),
    `${rule.failedCount}`,
  ]);

  return <Table withHeader rows={[headerRow, ...failedRuleRows]} />;
};

RulesTable.propTypes = {
  rules: propTypes.array,
};

export default RulesTable;
