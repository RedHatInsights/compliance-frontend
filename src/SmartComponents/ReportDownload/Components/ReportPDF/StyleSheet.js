import { StyleSheet } from '@react-pdf/renderer';

// Supported style properties see: https://react-pdf.org/styling
export default StyleSheet.create({
  // TODO: bold rendering in PDF does not appear to work
  bold: {
    fontWeight: 'bold',
  },
  italic: {
    fontStyle: 'italic',
  },

  metaTitle: {
    color: '#C9190B',
    marginBottom: '20px',
  },

  userNotes: {
    backgroundColor: '#F0F0F0',
    padding: '8px',
  },
  userNotesTitle: {
    marginBottom: '4px',
  },

  sectionTitle: {
    color: '#C9190B',
    fontWeight: 'bold',
    fontSize: 14,
  },

  subSectionTitle: {
    color: '#C9190B',
    fontWeight: 'bold',
    fontSize: 10,
    marginBottom: '10pt',
  },
});
