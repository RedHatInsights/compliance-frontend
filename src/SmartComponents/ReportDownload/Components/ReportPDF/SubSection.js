import React from 'react';
import propTypes from 'prop-types';
import { Text, View } from '@react-pdf/renderer';
import styles from './StyleSheet';

const SubSection = ({ children, metaTitle, title }) => {
  return (
    <View
      style={{
        marginBottom: '20pt',
      }}
    >
      <View
        style={{
          justifyContent: 'start',
        }}
      >
        <Text
          style={{
            ...styles.subSectionTitle,
            width: '70%',
          }}
        >
          {title}
        </Text>
        <Text>{metaTitle}</Text>
      </View>

      {children}
    </View>
  );
};

SubSection.propTypes = {
  children: propTypes.node,
  metaTitle: propTypes.string,
  title: propTypes.string,
};

export default SubSection;
