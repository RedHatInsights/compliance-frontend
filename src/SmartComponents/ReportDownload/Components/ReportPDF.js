import React, { Fragment } from 'react';
import propTypes from 'prop-types';
import { Text, View } from '@react-pdf/renderer';
// eslint-disable-next-line rulesdir/disallow-fec-relative-imports
import {
  Panel,
  Table,
  Column,
  Section,
} from '@redhat-cloud-services/frontend-components-pdf-generator';
import { DateFormat } from '@redhat-cloud-services/frontend-components/DateFormat';
import { fixedPercentage } from 'Utilities/TextHelper';
import styles from './ReportPDF/StyleSheet';
import SystemsTable from './ReportPDF/SystemsTable';
import UnsupportedSystemsTable from './ReportPDF/UnsupportedSystemsTable';
import NonReportedSystemsTable from './ReportPDF/NonReportedSystemsTable';
import RulesTable from './ReportPDF/RulesTable';
import PanelItem from './ReportPDF/PanelItem';
import SubSection from './ReportPDF/SubSection';
import ComplianceChart from './ReportPDF/ComplianceChart';

const ReportPDF = ({ data, ssgFinder }) => {
  const {
    totalHostCount,
    compliantSystems,
    compliantSystemCount,
    nonCompliantSystems,
    nonCompliantSystemCount,
    topTenFailedRules,
    unsupportedSystems,
    unsupportedSystemCount,
    nonReportingSystems,
    nonReportingSystemCount,
    userNotes,
    policy,
  } = data;
  const { testResultHostCount = 0, compliantHostCount = 0 } = policy;
  const percentCompliant =
    testResultHostCount != 0
      ? fixedPercentage(Math.floor(100 * (compliantHostCount / totalHostCount)))
      : 0;

  return (
    <Fragment>
      <Text style={styles.subSectionTitle}>{`Report prepared ${
        DateFormat({ date: new Date(), type: 'exact' }).props.children
      }`}</Text>

      {userNotes && (
        <View style={styles.userNotes}>
          <View style={styles.userNotesTitle}>
            <Text>User notes</Text>
          </View>
          <View>
            <Text>{userNotes}</Text>
          </View>
        </View>
      )}

      <Section
        title="Policy Details"
        titleProps={{
          style: styles.sectionTitle,
        }}
      >
        <Column style={{ width: '150px' }}>
          <Table
            // TODO: correct left side styling
            rows={[
              ['Policy type', policy.policyType],
              ['Operating system', `RHEL ${policy.osMajorVersion}`],
              ['Compliance threshold', `${policy.complianceThreshold}%`],
              ['Business Objective', policy.businessObjective || '--'],
            ]}
          />
        </Column>
        <Column>
          <ComplianceChart
            policy={{
              ...policy,
              percentCompliant,
            }}
            {...{
              compliantSystemCount,
              nonCompliantSystemCount,
              unsupportedSystemCount,
              nonReportingSystemCount,
            }}
          />
        </Column>
      </Section>

      <Section
        title="Systems"
        withColumn={false}
        titleProps={{
          style: styles.sectionTitle,
        }}
      >
        <Panel withColumn={false} style={{ marginBottom: '20px' }}>
          <PanelItem title="Non-compliant systems">
            {nonCompliantSystemCount}
          </PanelItem>

          {unsupportedSystemCount ? (
            <PanelItem title="Systems with unsupported configuration">
              {unsupportedSystemCount}
            </PanelItem>
          ) : null}

          {nonReportingSystemCount ? (
            <PanelItem title="Systems never reported">
              {nonReportingSystemCount}
            </PanelItem>
          ) : null}

          <PanelItem title="Compliant systems">
            {compliantSystemCount}
          </PanelItem>
        </Panel>

        {nonCompliantSystems && nonCompliantSystemCount ? (
          <SubSection title="Non-compliant systems">
            <SystemsTable systems={nonCompliantSystems} />
          </SubSection>
        ) : null}

        {unsupportedSystems && unsupportedSystemCount ? (
          <SubSection title="Systems with unsupported configuration">
            <UnsupportedSystemsTable
              systems={unsupportedSystems}
              ssgFinder={ssgFinder}
            />
          </SubSection>
        ) : null}

        {nonReportingSystems && nonReportingSystemCount ? (
          <SubSection title="Systems never reported">
            <NonReportedSystemsTable systems={nonReportingSystems} />
          </SubSection>
        ) : null}

        {compliantSystems && compliantSystemCount ? (
          <SubSection title="Compliant systems">
            <SystemsTable systems={compliantSystems} />
          </SubSection>
        ) : null}
      </Section>

      {topTenFailedRules ? (
        <Section
          title="Rules"
          withColumn={false}
          titleProps={{
            style: styles.sectionTitle,
          }}
        >
          <SubSection title="Top failed rules">
            <RulesTable rules={topTenFailedRules} />
          </SubSection>
        </Section>
      ) : null}
    </Fragment>
  );
};

ReportPDF.propTypes = {
  data: propTypes.object,
  ssgFinder: propTypes.func,
};

export default ReportPDF;
