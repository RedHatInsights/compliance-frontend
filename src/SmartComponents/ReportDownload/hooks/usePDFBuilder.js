import React from 'react';
import ReportPDF from '../Components/ReportPDF';

// (Eventually) responsible for rendering sections and splitting into pages
const usePDFBuilder = (policy) => async (data, ssgFinder) =>
  [
    <ReportPDF
      key="pdf-page-1"
      data={{ ...data, policy }}
      ssgFinder={ssgFinder}
    />,
  ];

export default usePDFBuilder;
