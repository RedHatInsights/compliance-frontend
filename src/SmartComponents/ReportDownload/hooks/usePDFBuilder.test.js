import { renderHook } from '@testing-library/react';
import usePDFBuilder from './usePDFBuilder';

import ReportPDF from '../Components/ReportPDF';
jest.mock('../Components/ReportPDF', () => jest.fn());
ReportPDF.mockImplementation(() => <div>PDF</div>);

describe('usePDFBuilder', () => {
  it('returns a build pages function', () => {
    const {
      result: { current: buildPages },
    } = renderHook(() => usePDFBuilder());
    expect(buildPages).toMatchSnapshot();
  });

  describe('buildPages', () => {
    it('returns and array of pages', async () => {
      const {
        result: { current: buildPages },
      } = renderHook(() => usePDFBuilder());
      expect(await buildPages({})).toMatchSnapshot();
    });
  });
});
