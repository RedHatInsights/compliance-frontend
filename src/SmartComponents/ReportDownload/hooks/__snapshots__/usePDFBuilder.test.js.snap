// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePDFBuilder buildPages returns and array of pages 1`] = `
[
  <mockConstructor
    data={
      {
        "policy": undefined,
      }
    }
  />,
]
`;

exports[`usePDFBuilder returns a build pages function 1`] = `[Function]`;
